apps/prairielearn/src/ee/lib/billing/plans.sql

```diff
@@ -1,24 +1,95 @@
--- BLOCK select_plan_grants_for_institution
+-- BLOCK select_plan_grants_for_context
 SELECT
-  *
+  pg.*
 FROM
-  plan_grants
+  plan_grants AS pg
 WHERE
-  institution_id = $institution_id
-  AND course_instance_id IS NULL
-  AND user_id IS NULL
-  and enrollment_id IS NULL;
+  -- Institution plan grant
+  (
+    $institution_id::bigint IS NOT NULL
+    AND $course_instance_id::bigint IS NULL
+    AND $enrollment_id::bigint IS NULL
+    AND $user_id::bigint IS NULL
+    AND institution_id = $institution_id
+    AND course_instance_id IS NULL
+    AND enrollment_id IS NULL
+    AND user_id IS NULL
+  )
+  -- Course instance plan grant
+  OR (
+    $institution_id::bigint IS NOT NULL
+    AND $course_instance_id::bigint IS NOT NULL
+    AND $enrollment_id::bigint IS NULL
+    AND $user_id::bigint IS NULL
+    AND institution_id = $institution_id
+    AND course_instance_id = $course_instance_id
+    AND enrollment_id IS NULL
+    AND user_id IS NULL
+  )
+  -- Enrollment plan grant
+  OR (
+    $institution_id::bigint IS NOT NULL
+    AND $course_instance_id::bigint IS NOT NULL
+    AND $enrollment_id::bigint IS NOT NULL
+    AND $user_id::bigint IS NULL
+    AND institution_id = $institution_id
+    AND course_instance_id = $course_instance_id
+    AND enrollment_id = $enrollment_id
+    AND user_id IS NULL
+  )
+  -- User plan grant
+  OR (
+    $institution_id::bigint IS NULL
+    AND $course_instance_id::bigint IS NULL
+    AND $enrollment_id::bigint IS NULL
+    AND $user_id::bigint IS NOT NULL
+    AND institution_id IS NULL
+    AND course_instance_id IS NULL
+    AND enrollment_id IS NULL
+    AND user_id = $user_id
+  );
 
--- BLOCK select_plan_grants_for_course_instance
+-- BLOCK select_plan_grants_for_context_recursive
```

Comment:
How does this query differ from the `select_plan_grants_for_context` query? I think they are functionally identical, given the comment above about the redundant `NOT NULL` checks?

Reply:
The tests prove that they are not functionally identical 🙂

The difference is that the non-recursive query will only match an exact context, which is hopefully made much clearer by the way I rewrote it in my comment above. The recursive version, however, will match partial contexts. That is, if I give the context `{institution_id: 1, course_instance_id: 1, user_id: 1}` and I have a plan grant with `{institution_id: null, course_instance_id: null, user_id: 1}`, the non-recursive one won't match this plan grant, but the recursive one will.

In retrospect, "recursive" might not be the best adjective here. Maybe... `select_plan_grants_for_partial_contexts`? Open to ideas on how to name this better!

Reply:
Yep, now that I thought about it for another minute I agree with you 😄. I do like `partial_contexts` better than `recursive`, but I'm fine with whatever name you prefer here. I understood the intent of the code, it's just on my fast initial read I didn't think enough about how the WHERE conditions were actually working!

Reply:
Renamed to "partial contexts" in b283082!